### PR TITLE
8333334: C2: Make result of `Node::dominates` more precise to enhance scalar replacement

### PR DESCRIPTION
**Pull request by MaxXSoft**
This patch changes the algorithm of `Node::dominates` to make the result more precise, and allows the iterators of `ConcurrentHashMap` to be scalar replaced.

The previous algorithm will return a conservative result when encountering a dead control flow, and only try the first two input paths of a multi-input Region node, which may prevent the scalar replacement in some cases.

For example, with G1 GC enabled, C2 generates GC barriers for `ConcurrentHashMap` iteration operations at some early phases, and then eliminates them in a later IGVN, but `LoadNode` is also idealized in the same IGVN. This causes `LoadNode::Ideal` to see some dead barrier control flows, and refuse to split some instance field loads through Phi due to the conservative result of `Node::dominates`, and thus the scalar replacement can not be applied to iterators in the later macro elimination phase.

This patch allows `Node::dominates` to try other paths of the last multi-input Region node when the first path is dead, and makes `ConcurrentHashMap` iteration ~30% faster:

```
Benchmark                            (nkeys)  Mode  Cnt        Score       Error  Units
Maps.testConcurrentHashMapIterators    10000  avgt   15   414099.085 ± 33230.945  ns/op   # baseline
Maps.testConcurrentHashMapIterators    10000  avgt   15   315490.281 ±  3037.056  ns/op   # patch
```

Testing: tier1-4.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8333334](https://bugs.openjdk.org/browse/JDK-8333334): C2: Make result of `Node::dominates` more precise to enhance scalar replacement (**Enhancement** - P4)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) Review applies to [8ee38498](https://git.openjdk.org/jdk/pull/19496/files/8ee384988889d79a60700b5732255269be6c50ab)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19496/head:pull/19496` \
`$ git checkout pull/19496`

Update a local copy of the PR: \
`$ git checkout pull/19496` \
`$ git pull https://git.openjdk.org/jdk.git pull/19496/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19496`

View PR using the GUI difftool: \
`$ git pr show -t 19496`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19496.diff">https://git.openjdk.org/jdk/pull/19496.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19496#issuecomment-2141556384)

### CODE REVIEW

**Review by TobiHartmann — COMMENTED** [review]
Impressive results! I haven't looked at the change yet but here are a few questions / requests:
- Could you add a screenshot of the IR of the case you are describing?
- Wouldn't it help to add the LoadNode back to the IGVN worklist and wait for the dead path to be removed?
- Could you add an [IR framework](https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/compiler/lib/ir_framework/README.md) test that verifies that the optimization works as expected?

Thanks,
Tobias

**Comment by MaxXSoft** [comment]
@TobiHartmann  Hi Tobias, thanks for your reply!

> Could you add a screenshot of the IR of the case you are describing?

Sure. Here's a simple example of iterating over the keys of `ConcurrentHashMap`:

```java
public long sumMapKeys() {
  long sum = 0;
  Enumeration it = map.keys();
  while (it.hasMoreElements()) {
    sum += (Long) it.nextElement();
  }
  return sum;
}
```

And here's what `-XX:+PrintEscapeAnalysis -XX:+PrintEliminateAllocations` says:

```
JavaObject(6) NoEscape(NoEscape) [ 183F 189F 205F 196F 191F 179F 186F 202F 208F 233F 637F 1069F [ 104 109 ]]     92  Allocate  === 76 6 69 8 1 (90 89 24 1 1 1 22 1 1 43 77 87 ) [[ 93 94 95 102 103 104 ]]  rawptr:NotNull ( int:>=0, java/lang/Object:NotNull *, bool, top, bool ) ConcurrentHashMap::keys @ bci:16 (line 2152) MyBenchmark::sumMapKeys @ bci:6 (line 105) !jvms: ConcurrentHashMap::keys @ bci:16 (line 2152) MyBenchmark::sumMapKeys @ bci:6 (line 105)
LocalVar(60) [ 92P [ 109 183b 189b 205b ]]    104  Proj  === 92  [[ 105 109 183 189 205 ]] #5 !jvms: ConcurrentHashMap::keys @ bci:16 (line 2152) MyBenchmark::sumMapKeys @ bci:6 (line 105)
LocalVar(103) [ 104 92P [ 196b 191b 179b 186b 202b 208b 233b 637b 1069b ]]    109  CheckCastPP  === 106 104  [[ 1801 1771 1754 1584 1503 1503 1490 1490 1466 1466 1451 1451 208 1393 1381 1381 179 179 186 186 208 196 191 191 196 202 202 228 297 233 233 1363 1363 1393 1321 1321 1311 637 637 648 648 998 988 1311 1301 477 477 487 487 497 497 569 1301 672 685 767 672 978 920 539 539 1069 1069 557 557 1128 569 631 1061 685 631 ]]  #java/util/concurrent/ConcurrentHashMap$KeyIterator (java/util/Iterator,java/util/Enumeration):NotNull:exact *  Oop:java/util/concurrent/ConcurrentHashMap$KeyIterator (java/util/Iterator,java/util/Enumeration):NotNull:exact * !jvms: ConcurrentHashMap::keys @ bci:16 (line 2152) MyBenchmark::sumMapKeys @ bci:6 (line 105)

NotScalar (Field load)  109  CheckCastPP  === 106 104  [[ 1801 1771 1754 1584 1503 1503 1490 1490 1128 569 1451 1451 208 1393 1381 1381 672 685 1069 1069 208 196 191 191 196 978 685 631 297 767 672 569 1301 1393 1321 1321 1311 557 557 631 1061 998 988 1311 1301 477 477 487 487 497 497 ]]  #java/util/concurrent/ConcurrentHashMap$KeyIterator (java/util/Iterator,java/util/Enumeration):NotNull:exact *,iid=92  Oop:java/util/concurrent/ConcurrentHashMap$KeyIterator (java/util/Iterator,java/util/Enumeration):NotNull:exact *,iid=92 !jvms: ConcurrentHashMap::keys @ bci:16 (line 2152) MyBenchmark::sumMapKeys @ bci:6 (line 105)
  >>>>  2186  LoadI  === _ 1973 191  [[ 2185 ]]  @java/util/concurrent/ConcurrentHashMap$Traverser+12 *, name=index, idx=9; #int !orig=[2178],[2171],[1373] !jvms: ConcurrentHashMap$Traverser::advance @ bci:51 (line 3369) ConcurrentHashMap$KeyIterator::next @ bci:28 (line 3468) ConcurrentHashMap$KeyIterator::nextElement @ bci:1 (line 3472) MyBenchmark::sumMapKeys @ bci:21 (line 107)
```

It shows that scalar replacement is aborted due to field load 2186:

ir-diff

As we can see its memory is a Phi, and it should be split by `LoadNode::split_through_phi` if its address `109 CheckCastPP` dominates the control flow of the Phi node `1330 Region`.

The control node of `CheckCastPP` is `106 Proj`:

cast

And it does dominate 1330, although not that obvious:

cfg

But `Node::dominates` think it doesn't because of the dead control flow.

> Wouldn't it help to add the LoadNode back to the IGVN worklist and wait for the dead path to be removed?

I tried to revert the change of main algorithm of `Node::dominates`, and just add some code to add the LoadNode back to the worklist if we met dead path. It works, still makes the iteration ~30% faster:

```
Benchmark                                  (nkeys)  Mode  Cnt        Score       Error  Units
Maps.testConcurrentHashMapIterators          10000  avgt   15   312720.415 ±  3255.500  ns/op
```

Thanks for pointing this out. I updated this PR, and the latest commit is passing test tier1-4.

> Could you add an [IR framework](https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/compiler/lib/ir_framework/README.md) test that verifies that the optimization works as expected?

I'm still trying to reproduce this case without using `ConcurrentHashMap`, but haven't found a way yet. Can I add an IR test that depends on `ConcurrentHashMap`? (It might need to be updated if the implementation of `ConcurrentHashMap` changes, I guess.)

**Comment by MaxXSoft** [comment]
Hi, I added an IR test to verify the optimization, now this patch is ready for review.

**Review by TobiHartmann — COMMENTED** [review]
> I tried to revert the change of main algorithm of Node::dominates, and just add some code to add the LoadNode back to the worklist if we met dead path. It works, still makes the iteration ~30% faster

That's great. Are we sure that what we detect as "dead code" can always be removed by IGVN? Otherwise, we risk an endless loop of putting the LoadNode back on the worklist and finding the dead code again, right?

> I'm still trying to reproduce this case without using ConcurrentHashMap, but haven't found a way yet. Can I add an IR test that depends on ConcurrentHashMap? (It might need to be updated if the implementation of ConcurrentHashMap changes, I guess.)

I think it would be great to have a standalone test, did you try with `-XX:+StressIGVN` and or `-XX:+StressIncrementalInlining`?

**Comment by TobiHartmann** [comment]
I executed some quick testing and I see a failure:

```
One or more @IR rules failed:

Failed IR Rules (1) of Methods (1)
----------------------------------
1) Method "private int compiler.c2.irTests.scalarReplacement.ScalarReplacementWithGCBarrierTests.testScalarReplacementWithGCBarrier(compiler.c2.irTests.scalarReplacement.ScalarReplacementWithGCBarrierTests$List)" - [Failed IR rules: 1]:
   * @IR rule 2: "@compiler.lib.ir_framework.IR(phase={INCREMENTAL_BOXING_INLINE}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={"_#ALLOC#_", "2"}, failOn={}, applyIfPlatform={}, applyIfPlatformOr={}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={}, applyIfAnd={}, applyIfNot={})"
     > Phase "Incremental Boxing Inline":
       - counts: Graph contains wrong number of nodes:
         * Constraint 1: "(\\d+(\\s){2}(Allocate\\b.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 1 = 2 [given]
             - Matched node:
               * 62  Allocate  === 30 6 7 8 1 (60 59 41 1 1 1 38 1 38 1 ) [[ 63 64 65 72 73 74 ]]  rawptr:NotNull ( int:>=0, java/lang/Object:NotNull *, bool, top, bool ) ScalarReplacementWithGCBarrierTests$List::iter @ bci:0 (line 48) ScalarReplacementWithGCBarrierTests::testScalarReplacementWithGCBarrier @ bci:1 (line 106) !jvms: ScalarReplacementWithGCBarrierTests$List::iter @ bci:0 (line 48) ScalarReplacementWithGCBarrierTests::testScalarReplacementWithGCBarrier @ bci:1 (line 106)
```
[log.txt](https://github.com/user-attachments/files/15814400/log.txt)

**Comment by TobiHartmann** [comment]
> For the latter case, the control node chain should be replaced by top and then eliminated (otherwise there must be something suboptimal).

Right, my concern was that not all control nodes would be folded but we have asserts that would catch such an endless loop during IGVN.

> I have already added a standalone IR test that does not rely on ConcurrentHashMap to this patch.

Great! Please have a look at the failure that I reported above.

**Comment by MaxXSoft** [comment]
> Great! Please have a look at the failure that I reported above.

I noticed that the IR test log shows there is only one Allocate node in IR, while I can not reproduce this with my JDK clone, even after rebasing to the latest commit and using the same VM options.

The IR test that I added asserts there are two allocations after boxing inlining, one for `Iter`, one for the `Integer` inside it. But in the IR dump, there's only one `Iter` allocation left, and all the LoadN/LoadI nodes of `Integer` come out from that allocation. I don't know which optimization performs this.

Am I missing something? Sorry for not familiar with the IR testing framework, maybe providing an option to match allocations at exact locations would help to build this test.

I'm still trying to get the test to work. BTW, could you observe the performance improvement in `ConcurrentHashMap` iteration after applying this patch to your JDK?

**Comment by TobiHartmann** [comment]
> Am I missing something? Sorry for not familiar with the IR testing framework, maybe providing an option to match allocations at exact locations would help to build this test.

I don't have time to look into this in detail right now but usually such issue come from a method not being inlined (in that case you need to enforce inlining via [@ForceInline](https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/compiler/lib/ir_framework/ForceInline.java) or via compile commands) or because an uncommon trap is added. You should be able to see that in the IR dump.

> BTW, could you observe the performance improvement in ConcurrentHashMap iteration after applying this patch to your JDK?

No, I didn't get a chance to try this out yet.

**Comment by MaxXSoft** [comment]
@TobiHartmann Hi!

Thanks for your previous suggestions. After several attempts, I can reproduce the failure after adding `@DontInline` for `List::iter`.

I have added `@ForceInline` for all relevant methods to ensure they are all inlined properly, and this patch is now ready for review.

File: test/hotspot/jtreg/compiler/c2/irTests/scalarReplacement/ScalarReplacementWithGCBarrierTests.java
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c2/irTests/scalarReplacement/ScalarReplacementWithGCBarrierTests.java:85]: This will fail if a different GC is selected via the command line. Your test needs a corresponding `@requires`.

Comment (MaxXSoft) [reply test/hotspot/jtreg/compiler/c2/irTests/scalarReplacement/ScalarReplacementWithGCBarrierTests.java:86]: Added `@requires vm.gc.G1`.

**Comment by TobiHartmann** [comment]
Thanks, test results look good now. I'll pass the detailed review to someone else.

File: src/hotspot/share/opto/memnode.hpp
Comment (eme64) [inline src/hotspot/share/opto/memnode.hpp:109]:
```suggestion
  static bool all_controls_dominate(Node* dom, Node* sub, bool* dead_code = nullptr);
```
We put the `*` next to the type. Putting it with the variable is old style - so you will see it but we try to change it whenever we touch a line.

File: src/hotspot/share/opto/memnode.cpp
Comment (eme64) [inline src/hotspot/share/opto/memnode.cpp:1677]:
```suggestion
  bool dead_code     = false;
  bool can_not_split = false;
```
With your version, would actually both variables be set to `false`, or could one be uninitialized? I just don't know enough about C++ here - but this makes me nervous ;)

File: src/hotspot/share/opto/memnode.cpp
Comment (eme64) [inline src/hotspot/share/opto/memnode.cpp:431]: This line is hard to read. And I'm not sure that `dummy_flag` is initialized. If not, it could most of the time happen to be `false`, but also randomly `true` sometimes. I would put the two variable definitions on separate lines.

File: src/hotspot/share/opto/memnode.cpp
Comment (eme64) [inline src/hotspot/share/opto/memnode.cpp:430]:
```suggestion
bool MemNode::all_controls_dominate(Node* dom, Node* sub, bool* dead_code) {
```
style

File: src/hotspot/share/opto/memnode.cpp
Comment (eme64) [reply src/hotspot/share/opto/memnode.cpp:431]: Also, this looks like a nasty hack with the pointer variable and the `dummy_flag`. Can we not solve this differently somehow?

File: src/hotspot/share/opto/memnode.cpp
Comment (eme64) [inline src/hotspot/share/opto/memnode.cpp:510]: A few comments about this line:
- You need curly braces for the body
- We do not do implicit null-checks: you should always do `dead_code != null`. But why are you even checking this here? In all other cases you would just set the `dead_code_flag`, and if `dead_code` then this would go to the `dummy_flag`, right?

**Review by eme64 — CHANGES_REQUESTED** [review]
Nice, looks like this is a worth-wile improvement!

I have some comments about style and software-engineering. Mainly:
- Generally return values passed via a argument is already a bit of a nasty solution - but sometimes it is the least nasty solution - so fine.
- But if you do pass a return value via argument, it should probably be a reference, not a pointer. But yours is not just a pointer, but an optional pointer.
- That leads you to the `dummy_flag` hack.
- How about this alternative: You create two `MemNode::all_controls_dominate` methods, one without the `bool& dead_code`, and one with. The one without simply creates a local variable and delegates to the implementation with the return-argument. Would that work?
- I would also give `dead_code` a more descriptive name. Currently it is not immediately clear what this flag is. For one, you should probably mention above `MemNode::all_controls_dominate` that it is a return-argument. Maybe call it `encountered_dead_code`.

Also: there are only 11 occurances of `all_controls_dominate`. Maybe you could have a `enum` return instead of a `bool`. It could have states `Dominate`, `Not_Dominate` and `Encountered_Dead_Code`. You would of course have to adjust all use-cases accordingly - but maybe we want to look at all use-cases anyway and make sure we handle the dead-code case correctly.

Side note:
`bool dummy_flag, &dead_code_flag = dead_code != nullptr ? *dead_code : dummy_flag;`

Lines like this make me really nervous. I'm not sure if the `dummy_flag` is initialized (maybe it doesn't matter, but still).

I'm not settled on any particular solution here, these are just my first reactions ;)

What do you think?

**Comment by MaxXSoft** [comment]
@eme64 Thanks for your patience and detailed comments!

I agree with you that returning values via arguments, especially pointer arguments, is not a good idea. I think returning a enum value would be an acceptable solution, so I've made the changes for methods `Node::dominates` and `MemNode::all_controls_dominate` to return `enum class DomResult` with values `Dominate`, `NotDominate` and `EncounteredDeadCode` as you suggested.

This solution does not contain any dummy flags or other hacks.

File: src/hotspot/share/opto/memnode.cpp
Comment (MaxXSoft) [reply src/hotspot/share/opto/memnode.cpp]: This line is even sillier than I thought once I understood your comment, please allow me to explain:

In this scope, `dead_code` is not the pointer argument of the method, it's just a local variable with the same name. I declared it here to accept the return value of `Node::dominates`, and if it's true, we should set the pointer argument to true to indicate that we have encountered a dead control flow.

This is a bad practice as the name of the variable can cause confusion. Sorry for writing this🤦, I've removed the relevant code in the latest commit.

File: test/hotspot/jtreg/compiler/c2/irTests/scalarReplacement/ScalarReplacementWithGCBarrierTests.java
Comment (dafedafe) [inline test/hotspot/jtreg/compiler/c2/irTests/scalarReplacement/ScalarReplacementWithGCBarrierTests.java:107]: Your test checks for the number of allocations to return to one. However in your description and comments you talk about load nodes that aren't folded. What about adding another IR test to check the number of loads for completeness?

**Review by dafedafe — COMMENTED** [review]
Nice improvement @MaxXSoft!

Quick note:  don't you want to add your copyright notice line to the files you've modified?

File: test/hotspot/jtreg/compiler/c2/irTests/scalarReplacement/ScalarReplacementWithGCBarrierTests.java
Comment (MaxXSoft) [reply test/hotspot/jtreg/compiler/c2/irTests/scalarReplacement/ScalarReplacementWithGCBarrierTests.java:107]: I think it's hard to add an intuitive and meaningful IR test for load nodes in this test case.

To construct a case that containing both eliminable GC pre-barriers and allocations, and have them interfere with each other, I had to put several object field reads/writes into nested loops. This makes a large number of load nodes appearing in this test case after some optimization phases. Perhaps after several more optimizations, most of them will disappear and new loads will appear again. So it's difficult to tell which load we really care about without looking at the Ideal graph dump via IGV.

**Comment by MaxXSoft** [comment]
@dafedafe Thanks for your suggestion! I've added copyright to all modified files.

**Comment by MaxXSoft** [comment]
Hi all,

This patch has now passed all GHA tests and is ready for further reviews.

If there are any other suggestions for this PR, please let me know.

Thanks!

File: src/hotspot/share/opto/memnode.cpp
Comment (TobiHartmann) [inline src/hotspot/share/opto/memnode.cpp:1664]: `igvn` should be moved below to its only usage.

File: test/hotspot/jtreg/compiler/c2/irTests/scalarReplacement/ScalarReplacementWithGCBarrierTests.java
Comment (TobiHartmann) [reply test/hotspot/jtreg/compiler/c2/irTests/scalarReplacement/ScalarReplacementWithGCBarrierTests.java:107]: Right, I think the test is fine as is.

**Review by TobiHartmann — COMMENTED** [review]
This looks good to me but you need a second review.

Please add brackets around if/else branches that you modified.

**Review by TobiHartmann — APPROVED** [review]

**Review by chhagedorn — COMMENTED** [review]
Generally, the fix idea looks good to me, too!

As Tobias has already mentioned, you should add some braces for the existing one-liner if-statements, just to be safe.

I have a few comment.

File: src/hotspot/share/opto/memnode.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/memnode.cpp:433]: Please, fix style of this and other code you touched. Add `{}` for condition's body where it is missing.

File: src/hotspot/share/opto/memnode.cpp
Comment (vnkozlov) [reply src/hotspot/share/opto/memnode.cpp:431]: I agree with Christian's suggestion about wrapper method which checks result.

**Comment by MaxXSoft** [comment]
@TobiHartmann @chhagedorn @vnkozlov Thanks for the suggestions! I've updated this PR.

**Review by TobiHartmann — APPROVED** [review]
Still good.

**Review by chhagedorn — APPROVED** [review]
Otherwise, the update looks good, thanks!

File: src/hotspot/share/opto/memnode.cpp
Comment (chhagedorn) [inline src/hotspot/share/opto/memnode.cpp:1717]:
```suggestion
      // There is some dead code which eventually will be removed in IGVN. Once this is the case, we get an unambiguous 
      // dominance result. Push the node to the worklist again until the dead code is removed.
```

**Review by chhagedorn — APPROVED** [review]

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by MaxXSoft** [comment]
@TobiHartmann @chhagedorn @vnkozlov Thanks for the review!

/integrate

**Pull request closed without merging**
